src/lib/cpuset.lua

```diff
@@ -19,35 +19,78 @@ do
    end
 end
 
+local function parse_cpulist (cpus)
+   local ret = {}
+   for range in cpus:split(',') do
+      local lo, hi = range:match("^%s*([^%-]*)%s*-%s*([^%-%s]*)%s*$")
+      if lo == nil then lo = range:match("^%s*([^%-]*)%s*$") end
+      assert(lo ~= nil, 'invalid range: '..range)
+      lo = assert(tonumber(lo), 'invalid range begin: '..lo)
+      assert(lo == math.floor(lo), 'invalid range begin: '..lo)
+      if hi ~= nil then
+         hi = assert(tonumber(hi), 'invalid range end: '..hi)
+         assert(hi == math.floor(hi), 'invalid range end: '..hi)
+         assert(lo < hi, 'invalid range: '..range)
+      else
+         hi = lo
+      end
+      for cpu=lo,hi do table.insert(ret, cpu) end
+   end
+   return ret
+end
+
+local function parse_cpulist_from_file (path)
+   local fd = assert(io.open(path))
+   if not fd then return {} end
+   local ret = parse_cpulist(fd:read("*all"))
+   fd:close()
+   return ret
+end
+
+local function available_cpus (node)
+   local function set (t)
+      local ret = {}
+      for _,v in pairs(t) do ret[tostring(v)] = true end
+      return ret
+   end
+   local function cpus_in_node (node)
+      local node_path = '/sys/devices/system/node/node'..node
+      return set(parse_cpulist_from_file(node_path..'/cpulist'))
+   end
+   local function isolated_cpus ()
+      return set(parse_cpulist_from_file('/sys/devices/system/cpu/isolated'))
+   end
+   local function substract (s, t)
+      local ret = {}
+      for k,_ in pairs(s) do
+         if not t[k] then table.insert(ret, k) end
+      end
+      table.sort(ret)
+      return ret
+   end
+   return substract(cpus_in_node(node), isolated_cpus())
```

Comment:
As a FIXME I think this should be unioned with the current CPU affinity (as returned by sched_getaffinity), to allow for users to run the manager with "taskset".  However, not a blocker!